### PR TITLE
Added sample interator api

### PR DESCRIPTION
**Pull request by JCash**
Fixes https://github.com/Celtoys/Remotery/issues/190

### CODE REVIEW

File: lib/Remotery.c [hunk 1]
```diff
@@ -4309,22 +4309,12 @@ static void Server_Update(Server* server)
 
 #define SAMPLE_NAME_LEN 128
 
-typedef enum SampleType
```
Comment (JCash) [inline lib/Remotery.c:4312]: Moved to the header, and renamed to `rmtSampleType`

File: lib/Remotery.c [hunk 2]
```diff
@@ -6115,6 +6105,11 @@ static rmtError Remotery_SendSampleTreeMessage(Remotery* rmt, Message* message)
     error = bin_SampleTree(bin_buf, sample_tree);
     rmt_EndCPUSample();
 
+    if (g_Settings.sampletree_handler != NULL)
+    {
+        g_Settings.sampletree_handler(rmt, g_Settings.sampletree_context, sample_tree);
+    }
```
Comment (JCash) [inline lib/Remotery.c:6111]: I had to put it before the `FreeSamples()` as the sample pointers would be invalid after that call.

File: lib/Remotery.c [hunk 3]
```diff
@@ -8820,4 +8817,92 @@ RMT_API void _rmt_EndMetalSample(void)
...
+        iter->initial = 0;
+    }
+    else
+    {
+        if (iter->sample != NULL)
+            iter->sample = iter->sample->next_sibling;
+    }
+
+    return iter->sample != NULL ? RMT_TRUE : RMT_FALSE;
+}
```
Comment (JCash) [inline lib/Remotery.c:8843]: Following the pattern from this gist:
https://gist.github.com/pervognsen/744066869da509a15c4e08fc9408207e

File: lib/Remotery.c [hunk 4]
```diff
@@ -8820,4 +8817,92 @@ RMT_API void _rmt_EndMetalSample(void)
...
+// Sample tree accessors
+RMT_API const char* _rmt_SampleTreeGetThreadName(rmtMsgSampleTree* sample_tree)
+{
+    return sample_tree->threadName;
+}
+
+RMT_API rmtSample* _rmt_SampleTreeGetRootSample(rmtMsgSampleTree* sample_tree)
+{
+    return sample_tree->rootSample;
+}
```
Comment (JCash) [inline lib/Remotery.c:8854]: At first, I passed the `threadName` explicitly to the callback function, but later thought that passing the opaque message pointer fits the api a bit better:
* Easier to add more accessor functions later
* Matches the Sample accessor api used in the callback api

File: lib/Remotery.c [hunk 5]
```diff
@@ -8820,4 +8817,92 @@ RMT_API void _rmt_EndMetalSample(void)
...
+// Sample accessors
+RMT_API const char* _rmt_SampleGetName(Remotery* rmt, rmtSample* sample)
+{
+    const char* name = StringTable_Find(rmt->string_table, sample->name_hash);
+    if (name == NULL)
+    {
+        return "null";
+    }
+    return name;
+}
```
Comment (JCash) [inline lib/Remotery.c:8865]: Needs the Remotery pointer, so I explicitly pass that in the callback.

File: lib/Remotery.c [hunk 6]
```diff
@@ -8820,4 +8817,92 @@ RMT_API void _rmt_EndMetalSample(void)
...
+
+RMT_API rmtU64 _rmt_SampleGetTime(rmtSample* sample)
+{
+    return sample->us_length;
+}
+
+RMT_API rmtU64 _rmt_SampleGetSelfTime(rmtSample* sample)
+{
+    return (rmtU64)maxS64(sample->us_length - sample->us_sampled_length, 0);
```
Comment (JCash) [inline lib/Remotery.c:8884]: I think this is correct, as it's the same as is sent over the socket.

File: lib/Remotery.h [hunk 7]
```diff
@@ -311,6 +316,26 @@ typedef enum rmtSampleFlags
...
 
+typedef enum rmtSampleType
+{
+    RMT_SampleType_CPU,
+    RMT_SampleType_CUDA,
+    RMT_SampleType_D3D11,
+    RMT_SampleType_OpenGL,
+    RMT_SampleType_Metal,
+    RMT_SampleType_Count,
+} rmtSampleType;
```
Comment (JCash) [inline lib/Remotery.h:327]: Given the other public interfaces in the header, I prefixed it with RMT_.

Might want to be upper case, to match style of `rmtError` enum?

File: lib/Remotery.h [hunk 8]
```diff
@@ -363,6 +388,7 @@ typedef void* (*rmtMallocPtr)(void* mm_context, rmtU32 size);
 typedef void* (*rmtReallocPtr)(void* mm_context, void* ptr, rmtU32 size);
 typedef void (*rmtFreePtr)(void* mm_context, void* ptr);
 typedef void (*rmtInputHandlerPtr)(const char* text, void* context);
+typedef void (*rmtSampleTreeHandlerPtr)(Remotery* rmt, void* cbk_context, rmtMsgSampleTree* sample_tree);
```
Comment (JCash) [inline lib/Remotery.h:391]: The callback type and its arguments.

File: lib/Remotery.h [hunk 9]
```diff
@@ -661,6 +691,23 @@ RMT_API void _rmt_BeginMetalSample(rmtPStr name, rmtU32* hash_cache);
...
+RMT_API rmtSample*          _rmt_SampleTreeGetRootSample(rmtMsgSampleTree* sample_tree);
+
+// Sample accessors
+RMT_API const char*         _rmt_SampleGetName(Remotery* rmt, rmtSample* sample);
+RMT_API rmtU32              _rmt_SampleGetNameHash(rmtSample* sample);
+RMT_API rmtU32              _rmt_SampleGetCallCount(rmtSample* sample);
+RMT_API rmtU64              _rmt_SampleGetTime(rmtSample* sample);
+RMT_API rmtU64              _rmt_SampleGetSelfTime(rmtSample* sample);
+RMT_API void                _rmt_SampleGetColour(rmtSample* sample, rmtU8* r, rmtU8* g, rmtU8* b);
+RMT_API rmtSampleType       _rmt_SampleGetType(rmtSample* sample);
```
Comment (JCash) [inline lib/Remotery.h:709]: Currently, they all are "private", with the "_" prefix.

I guess they should also be wrapped with the "RMT_OPTIONAL" define.
Should it also be contained withing a new tag, e.g. `RMT_ENABLED_ITERATORS` ? Might be a bit overkill?

File: sample/dump.c [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+void dumpTree(Remotery* rmt, void* ctx, rmtMsgSampleTree* sample_tree)
+{
+    rmtSample* root = _rmt_SampleTreeGetRootSample(sample_tree);
+    const char* thread_name = _rmt_SampleTreeGetThreadName(sample_tree);
+
+    printf("// ********************   DUMP TREE: %s   ************************\n", thread_name);
+
+    printTree(rmt, root, 0);
+}
```
Comment (JCash) [inline sample/dump.c:75]: Example callback

File: sample/dump.c [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+void printTree(Remotery* rmt, rmtSample* sample, int indent)
+{
+    printSample(rmt, sample, indent);
+
+    rmtSampleIterator iter = _rmt_IterateChildren(sample);
+    while (_rmt_IterateNext(&iter)) {
+        printTree(rmt, iter.sample, indent+1);
+    }
+}
```
Comment (JCash) [inline sample/dump.c:65]: Example of iterator

File: lib/Remotery.h [hunk 12]
```diff
@@ -661,6 +691,23 @@ RMT_API void _rmt_BeginMetalSample(rmtPStr name, rmtU32* hash_cache);
 RMT_API void _rmt_EndMetalSample(void);
 #endif
 
+// Iterator
```
Comment (dwilliamson) [inline lib/Remotery.h:694]: Does this API compile and link when `RMT_ENABLED=0`? Prefix-underscores shouldn't be used here.

As an example:

```cpp
RMT_API void _rmt_LogText(rmtPStr text);
```

This is the function signature that is part of the Private interface and this is the macro which gets compiled-out when `RMT_ENABLED=0`:

```cpp
#define rmt_LogText(text)                                                           \
    RMT_OPTIONAL(RMT_ENABLED, _rmt_LogText(text))
```

**Comment by dwilliamson** [comment]
Really great stuff.

I started reviewing this before you left your comments so I'll address them now.

File: lib/Remotery.c (hunk 4 above)
Comment (dwilliamson) [reply lib/Remotery.c:8854]: Hmmm... yes... that's a good counter argument. I'll edit my comment.

File: lib/Remotery.c [hunk 13]
```diff
@@ -6115,6 +6105,11 @@ static rmtError Remotery_SendSampleTreeMessage(Remotery* rmt, Message* message)
     error = bin_SampleTree(bin_buf, sample_tree);
     rmt_EndCPUSample();
 
+    if (g_Settings.sampletree_handler != NULL)
+    {
+        g_Settings.sampletree_handler(rmt, g_Settings.sampletree_context, sample_tree);
```
[the remark this answers is not in the document]
Comment (dwilliamson) [reply lib/Remotery.c:6110]: Leave as is but could you rename to `rmtSampleTree`?

File: lib/Remotery.c (hunk 5 above)
Comment (dwilliamson) [reply lib/Remotery.c:8865]: You should use `g_Remotery` in the implementation instead.

File: lib/Remotery.h (hunk 7 above)
Comment (dwilliamson) [reply lib/Remotery.h:327]: This is fine. I wanted to remove all error-messages-as-enums, anyway and replace them with 4 or 5 broader error enums that give you optional message retrieval.

File: lib/Remotery.h (hunk 9 above)
Comment (dwilliamson) [reply lib/Remotery.h:709]: Don't worry about adding a `RMT_ENABLED_ITERATORS` - just `RMT_ENABLED` will do.

File: lib/Remotery.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (JCash) [reply lib/Remotery.c:6110]: Sure. That's actually exactly what I had before :)

File: lib/Remotery.c (hunk 5 above)
Comment (JCash) [reply lib/Remotery.c:8865]: I was thinking that too, but I was perhaps trying to remove the need to use a global variable.

The current setup in the PR was to address the initialization problem:


The problem with the user context and the remotery pointer is a catch 22:
```
// Setting up settings before creating the context
Remotery *rmt;
rmtSettings* settings = rmt_Settings();
settings->sampletree_handler = dumpTree;
settings->sampletree_context = rmt; // Not valid yet
rmt_CreateGlobalInstance(&rmt);
```

An alternative, albeit not very clean code flow either:
```
// Setup some setting variables before creating the Remotery instance...
Remotery *rmt;
rmtSettings* settings = rmt_Settings();
settings->port = port;
rmt_CreateGlobalInstance(&rmt);

// ... and set some settings after
settings->sampletree_handler = dumpTree;
settings->sampletree_context = rmt; // Not valid yet
```

But, I have no examples of other users use case, and they can for sure create their own custom contexts etc.

So I'll go ahead with removing the Remotery pointer from the callback function.

File: lib/Remotery.h (hunk 12 above)
Comment (JCash) [reply lib/Remotery.h:694]: I have now added this, but I'm wondering if the use case is a bit different.
The client code will fail to compile on code like this if we simply omit the info:

```
error: expected expression
    const char* name = rmt_SampleGetName(ctx->rmt, sample);
```

I see two options:
1. Each such interface needs to produce "valid" values, which is counter intuitive when using the `RMT_OPTIONAL + RMT_ENABLED`
2. Letting the client control their code using `#ifdef RMT_ENABLED` explicitly around this particular piece of code.

Comment (dwilliamson) [reply lib/Remotery.h:731]: Yes, there's going to be a lot of errors if you implement some of them in terms of `RMT_OPTIONAL` as they need to return a value even when `RMT_ENABLED=0`.

One option is to return nothing and specify the output as an address parameter to the function call but that would then prevent you from initialising at the declaration point in C when multiple calls are involved.

Have you looked at using `RMT_OPTIONAL_RET`?

File: lib/Remotery.c (hunk 5 above)
Comment (dwilliamson) [reply lib/Remotery.c:8865]: Most of the internal code uses the global variable `g_Remotery`. The API explicitly avoids requesting it from the user. It even has a DLL API to cater for this where the use calls `rmt_SetGlobalInstance`.

Comment (JCash) [reply lib/Remotery.c:8865]: Ah, now I finally see what you mean. Fixing...

File: lib/Remotery.h [hunk 14]
```diff
@@ -358,11 +383,49 @@ typedef enum rmtSampleFlags
     RMT_OPTIONAL(RMT_ENABLED, _rmt_EndCPUSample())
 
 
+#define rmt_IterateChildren(iter, sample)                                           \
+    RMT_OPTIONAL(RMT_ENABLED, _rmt_IterateChildren(iter, sample))
```
Comment (JCash) [inline lib/Remotery.h:387]: Tweaked the api from
```
rmtSampleIterator iter = rmt_IterateChildren(sample);
```
to
```
rmtSampleIterator iter;
rmt_IterateChildren(&iter, sample);
```
in order to both make it more C89 compatible, and also make it work with RMT_OPTIONAL better.

File: lib/Remotery.h (hunk 12 above)
Comment (JCash) [reply lib/Remotery.h:731]: I updated the api now to use the RMT_OPTIONAL+RMT_OPTIONAL_RET and the dump.c example builds and runs in `-DRMT_ENABLED=0` as well.

**Comment by dwilliamson** [comment]
I love this... going to get some tea and make sure I understand the implications of adding this before merge.

Thanks so much.

**Review by dwilliamson — APPROVED** [review]

**Pull request merged**